tensorpipe/channel/cuda/channel.cc

```diff
@@ -0,0 +1,406 @@
+                     ptr,
+                     length,
+                     callback{std::move(callback)}]() mutable {
+    recvFromLoop_(std::move(descriptor), ptr, length, std::move(callback));
+  });
+}
+
+void Channel::Impl::recvFromLoop_(
+    TDescriptor descriptor,
+    void* ptr,
+    size_t length,
+    TRecvCallback callback) {
+  TP_DCHECK(loop_.inLoop());
+
+  const uint64_t sequenceNumber = nextTensorBeingReceived_++;
+  TP_VLOG(4) << "Channel " << id_ << " received a recv request (#"
+             << sequenceNumber << ")";
+  callback = [this, sequenceNumber, callback{std::move(callback)}](
+                 const Error& error) {
+    TP_VLOG(4) << "Channel " << id_ << " is calling a recv callback (#"
+               << sequenceNumber << ")";
+    callback(error);
+    TP_VLOG(4) << "Channel " << id_ << " done calling a recv callback (#"
+               << sequenceNumber << ")";
+  };
+
+  if (error_) {
+    callback(error_);
+    return;
+  }
+
+  proto::Descriptor pbDescriptor;
+  loadDescriptor(pbDescriptor, descriptor);
+
+  const cudaIpcMemHandle_t* remoteHandle =
+      reinterpret_cast<const cudaIpcMemHandle_t*>(
+          pbDescriptor.handle().c_str());
+
+  const cudaIpcEventHandle_t* startEvHandle =
+      reinterpret_cast<const cudaIpcEventHandle_t*>(
+          pbDescriptor.start_ev_handle().c_str());
+  const cudaIpcEventHandle_t* stopEvHandle =
+      reinterpret_cast<const cudaIpcEventHandle_t*>(
+          pbDescriptor.stop_ev_handle().c_str());
+
+  cudaEvent_t startEv;
+  TP_DCHECK_CUDA(cudaIpcOpenEventHandle(&startEv, *startEvHandle));
+  cudaEvent_t stopEv;
+  TP_DCHECK_CUDA(cudaIpcOpenEventHandle(&stopEv, *stopEvHandle));
+
+  TP_VLOG(6) << "Channel " << id_ << " is copying payload (#" << sequenceNumber
+             << ")";
+
+  TP_DCHECK_CUDA(cudaStreamWaitEvent(stream_, startEv, 0));
+
+  // Perform copy.
+  if (length > 0) {
+    void* remotePtr;
+    TP_DCHECK_CUDA(cudaIpcOpenMemHandle(
+        &remotePtr, *remoteHandle, cudaIpcMemLazyEnablePeerAccess));
+
+    TP_DCHECK_CUDA(cudaMemcpyAsync(ptr, remotePtr, length, cudaMemcpyDefault));
+    TP_DCHECK_CUDA(cudaIpcCloseMemHandle(remotePtr));
+  }
+  TP_DCHECK_CUDA(cudaEventRecord(stopEv));
+
+  TP_VLOG(6) << "Channel " << id_ << " done copying payload (#"
+             << sequenceNumber << ")";
+
+  // Let peer know we've completed the copy.
+  TP_VLOG(6) << "Channel " << id_ << " is writing notification (#"
+             << sequenceNumber << ")";
+  connection_->write(
+      nullptr, 0, lazyCallbackWrapper_([sequenceNumber](Impl& impl) {
+        TP_VLOG(6) << "Channel " << impl.id_ << " done writing notification (#"
+                   << sequenceNumber << ")";
+      }));
+
+  callback(error_);
+}
+
+void Channel::setId(std::string id) {
+  impl_->setId(std::move(id));
+}
+
+void Channel::Impl::setId(std::string id) {
+  TP_VLOG(4) << "Channel " << id_ << " was renamed to " << id;
+  // FIXME Should we defer this to the loop?
```

Comment:
We found out that the answer to this question is yes! If you look at other channels this FIXME has been addressed, could you copy it over?